storage-proofs-update/src/vanilla.rs

```diff
@@ -0,0 +1,1209 @@
+        );
+
+        info!(
+            "new replica path {:?}, len {}",
+            new_replica_path,
+            new_replica_path_metadata.len()
+        );
+        info!(
+            "sector key path {:?}, len {}",
+            sector_key_path,
+            sector_key_path_metadata.len()
+        );
+        info!(
+            "staged data path {:?}, len {}",
+            staged_data_path,
+            staged_data_path_metadata.len()
+        );
+
+        // Setup read-only mmaps for sector_key_path and staged_data_path inputs.
+        let sector_key_data = mmap_read(sector_key_path)?;
+        let staged_data = mmap_read(staged_data_path)?;
+
+        // Setup writable mmap for new_replica_path output.
+        let mut new_replica_data = mmap_write(new_replica_path)?;
+
+        // Re-open staged_data as Data (type)
+        let mut new_data = Data::from_path(staged_data_path.to_path_buf());
+        new_data.ensure_data()?;
+
+        // Generate tree_d over the staged_data.
+        let tree_d = StackedDrg::<TreeR, TreeDHasher>::build_binary_tree::<TreeDHasher>(
+            new_data.as_ref(),
+            tree_d_new_config,
+        )?;
+
+        let comm_d_new = tree_d.root();
+
+        let comm_r_old = <TreeRHasher as Hasher>::Function::hash2(&comm_c, &comm_r_last_old);
+        let phi = phi(&comm_d_new, &comm_r_old);
+
+        let end = staged_data_path_metadata.len() as u64;
+
+        // chunk_size is the number of Fr elements to process in parallel chunks.
+        let chunk_size: usize = std::cmp::min(base_tree_nodes_count, CHUNK_SIZE_MIN);
+
+        // data_block_size is the segment length that we're processing
+        // in Fr elements (i.e. chunk_size * sizeof(Fr)).
+        let data_block_size: usize = chunk_size * FR_SIZE;
+
+        // AND-mask which strips the partition-index `k` from a node-index.
+        let node_index_bit_len = nodes_count.trailing_zeros() as usize;
+        let partition_count = partition_count(nodes_count);
+        let partition_bit_len = partition_count.trailing_zeros() as usize;
+        let remove_k_from_node_index_mask = (1 << (node_index_bit_len - partition_bit_len)) - 1;
+
+        Vec::from_iter((0..end).step_by(data_block_size))
+            .into_par_iter()
+            .zip(new_replica_data.par_chunks_mut(data_block_size))
+            .try_for_each(|(chunk_index, replica_data)| -> Result<()> {
+                for i in (0..data_block_size as u64).step_by(FR_SIZE) {
+                    let input_index = (chunk_index as usize) + i as usize;
+                    let output_index = i as usize;
+
+                    // Get the `h` high bits from the node-index (sans partition-index).
+                    let node_index = input_index / FR_SIZE;
+                    let input_index_without_k = node_index & remove_k_from_node_index_mask;
+                    let high =
+                        input_index_without_k >> (node_index_bit_len - partition_bit_len - h);
```

Comment:
I don't think `parititon_bit_len` should be used here.
We want `h` high bits out of the node index, the partition count is unimportant to the encoding process.